master/internal/stream/trials.go

```diff
@@ -54,39 +53,48 @@ func (tm *TrialMsg) SeqNum() int64 {
 	return tm.Seq
 }
 
-func (tm *TrialMsg) UpsertMsg() *websocket.PreparedMessage {
+func (tm *TrialMsg) UpsertMsg(upsertFunc stream.UpsertFunc) interface{} {
 	wrapper := struct {
 		Trial *TrialMsg `json:"trial"`
 	}{tm}
+
+	if upsertFunc != nil {
+		return upsertFunc(tm)
+	}
 	return prepareMessageWithCache(wrapper, &tm.upsertCache)
 }
 
-func (tm *TrialMsg) DeleteMsg() *websocket.PreparedMessage {
+func (tm *TrialMsg) DeleteMsg(deleteFunc stream.DeleteFunc) interface{} {
 	deleted := strconv.FormatInt(int64(tm.ID), 10)
+
+	if deleteFunc != nil {
+		return deleteFunc(TrialsDeleteKey, deleted)
+	}
 	return newDeletedMsgWithCache(TrialsDeleteKey, deleted, &tm.deleteCache)
 }
 
 // TrialSubscriptionSpec is what a user submits to define a trial subscription.
 // determined:streamable
 type TrialSubscriptionSpec struct {
-	TrialIds      []int  `json:"trial_ids"`
-	ExperimentIds []int  `json:"experiment_ids"`
-	Since         int64  `json:"since"`
+	TrialIds      []int `json:"trial_ids"`
+	ExperimentIds []int `json:"experiment_ids"`
+	Since         int64 `json:"since"`
 }
 
+// TODO: refactor pls
```

Comment:
out of curiosity: why do you say this?

Reply:
this was just a reminder for me to come back and change preparedMessages to interfaces. It shouldn't be relevant anymore